packages/calling-stateful-client/src/TeamsCallAgentDeclarative.ts

```diff
@@ -0,0 +1,128 @@
+import { clearCallRelatedState, DeclarativeCallCommon, ProxyCallAgentCommon } from './CallAgentDeclarativeCommon';
+import { CallContext } from './CallContext';
+/* @conditional-compile-remove(teams-identity-support) */
+import { isTeamsCall, isTeamsCallAgent } from './Converter';
+import { InternalCallContext } from './InternalCallContext';
+/* @conditional-compile-remove(teams-identity-support) */
+import { teamsCallDeclaratify } from './TeamsCallDeclarative';
+
+/**
+ * @beta
+ * `DeclarativeTeamsCallAgent` extends and proxies the {@link @azure/communication-calling#TeamsCallAgent}
+ */
+export type DeclarativeTeamsCallAgent = TeamsCallAgent &
+  /* @conditional-compile-remove(one-to-n-calling) */ IncomingCallManagement;
+
+/**
+ * ProxyTeamsCallAgent proxies TeamsCallAgent and saves any returned state in the given context. It will subscribe to all state
+ * updates in the TeamsCallAgent and in the contained TeamsCalls and RemoteParticipants. When dispose is called it will
+ * unsubscribe from all state updates.
+ */
+class ProxyTeamsCallAgent extends ProxyCallAgentCommon implements ProxyHandler<DeclarativeTeamsCallAgent> {
+  /* @conditional-compile-remove(teams-identity-support) */
+  private _callAgent: TeamsCallAgent;
+
+  constructor(callAgent: TeamsCallAgent, context: CallContext, internalContext: InternalCallContext) {
+    super(context, internalContext);
+    /* @conditional-compile-remove(teams-identity-support) */
+    this._callAgent = callAgent;
+    this.subscribe();
+  }
+
+  private subscribe = (): void => {
+    /* @conditional-compile-remove(teams-identity-support) */
+    this._callAgent.on('callsUpdated', this.callsUpdated);
+    /* @conditional-compile-remove(teams-identity-support) */
+    this._callAgent.on('incomingCall', this.incomingCall);
+
+    /* @conditional-compile-remove(teams-identity-support) */
+    // There could be scenario that when ProxyTeamsCallAgent is created that the given CallAgent already has TeamsCalls. In this
+    // case we need to make sure to subscribe to those already existing Calls.
+    for (const call of this._callAgent.calls) {
+      this.addCall(call);
+    }
+  };
+
+  protected unsubscribe = (): void => {
+    /* @conditional-compile-remove(teams-identity-support) */
+    this._callAgent.off('callsUpdated', this.callsUpdated);
+    /* @conditional-compile-remove(teams-identity-support) */
+    this._callAgent.off('incomingCall', this.incomingCall);
+
+    this.unregisterSubscriber();
+  };
+
+  protected callDeclaratify(call: CallCommon, context: CallContext): DeclarativeCallCommon {
+    /* @conditional-compile-remove(teams-identity-support) */
+    if (isTeamsCall(call)) {
+      return teamsCallDeclaratify(call, context);
+    }
+    throw new Error('Not reachable code, DeclarativeTeamsCallAgent.callDeclaratify must be called with an TeamsCall.');
+  }
+
+  protected startCall(agent: CallAgentCommon, args: unknown[]): CallCommon {
```

Comment:
why are all these `unknown`? This will become really difficult in the future for developers to use these args correctly

Reply:
(e.g. why aren't these `args: Parameters<TeamsCallAgent['startCall']>`)

Reply:
Oh because of the runtime check `isTeamsCallAgent` I see.. let me think on this more

Reply:
These are args coming from parent class override parameters, from parent class, it's hard to determine args(or the type would be super complicated and unreadable) because they are generic one

Reply:
Update:
New typescript support to do that only in child class, so I can actually type the parameter in child class but still keep parent class the same